core/packages/contracts/src/UpgradeProxy.sol

```diff
@@ -0,0 +1,56 @@
+// SPDX-License-Identifier: Apache-2.0
+pragma solidity ^0.8.19;
+
+import {AccessControl} from "openzeppelin/access/AccessControl.sol";
+import {IVault} from "./IVault.sol";
+import {IUpgradeTask} from "./IUpgradeTask.sol";
+import {ParaID} from "./Types.sol";
+
+contract UpgradeProxy is AccessControl {
+    bytes32 public constant ADMIN_ROLE = keccak256("ADMIN_ROLE");
+    bytes32 public constant SENDER_ROLE = keccak256("SENDER_ROLE");
+
+    struct Message {
+        Action action;
+        bytes payload;
+    }
+
+    enum Action {
+        Upgrade
+    }
+
+    struct UpgradePayload {
+        address task;
+    }
+
+    error InvalidMessage();
+    error Unauthorized();
+    error UpgradeFailed();
+
+    // Parachain ID of BridgeHub
+    ParaID public immutable bridgeHubParaID;
+
+    constructor(ParaID _bridgeHubParaID) {
+        _grantRole(ADMIN_ROLE, msg.sender);
+        _setRoleAdmin(SENDER_ROLE, ADMIN_ROLE);
+        bridgeHubParaID = _bridgeHubParaID;
+    }
+
+    function handle(ParaID origin, bytes calldata message) external onlyRole(SENDER_ROLE) {
+        if (origin != bridgeHubParaID) {
+            revert Unauthorized();
+        }
+
+        Message memory decoded = abi.decode(message, (Message));
+        if (decoded.action != Action.Upgrade) {
+            revert InvalidMessage();
+        }
+
+        UpgradePayload memory payload = abi.decode(decoded.payload, (UpgradePayload));
+
+        (bool success,) = payload.task.delegatecall(abi.encodeCall(IUpgradeTask.run, ()));
```

Comment:
may not be the scope of this PR,  just curious is the `IUpgradeTask` here meant only for Parameter-change like fee or reward from governance? For senarios like bug fixing or introduce new features do we need some [proxy pattern](https://docs.openzeppelin.com/upgrades-plugins/1.x/proxies) instead?

Reply:
Yeah that's exactly correct :+1:

Reply:
Though to add, there are two stateful contracts which we don't plan on replacing ever:
* Vault.sol: Controls locked up ETH on behalf of parachains/users
* Tokenvault.sol: Controls locked up ERC20 tokens on behalf of users

The idea is these contracts are very simple, and don't much logic that would need upgrading or replacing.

Our `NativeTokens` contract acts as a shell over `TokenVault`, implementing all the messy business logic of interacting with users and the bridge. That's the kind of logic which we would expect to be upgraded.

So when we deploy a new NativeTokens contract, we would allow the replacement control to access the existing tokenVault.